examples/usercall-extension-bind/runner/src/main.rs

```diff
@@ -0,0 +1,301 @@
+            self.buf.resize(reqlen, 0);
+        }
+        let read = self.stream.read(&mut self.buf[self.position..])?;
+        if read == 0 {
+            return Err(Error::new(ErrorKind::UnexpectedEof, "Missing data in proxy protocol"))
+        }
+        self.position += read;
+        if self.position >= reqlen {
+            Ok(())
+        } else {
+            Err(Error::from_raw_os_error(libc::EWOULDBLOCK))
+        }
+    }
+
+    fn sig(&self) -> &[u8] {
+        &self.buf[..SIZE_HEADER_SIG]
+    }
+
+    fn advance_stream(&mut self) -> IoResult<()> {
+        use byteorder::{NetworkEndian, ReadBytesExt};
+        self.read_upto(SIZE_HEADER)?;
+        let vercmd = self.buf[SIZE_HEADER_SIG + 0];
+        let family = self.buf[SIZE_HEADER_SIG + 1];
+        let addr_len = (&self.buf[(SIZE_HEADER_SIG + 2)..]).read_u16::<NetworkEndian>().unwrap();
+
+        if self.sig() != HEADER_SIG {
+            self.fuse("Protocol header signature mismatch")?
+        }
+
+        let min_len = match (vercmd, family) {
+            (V2CMD_LOCAL, FAMILY_UNSPEC) => 0,
+            (V2CMD_PROXY, FAMILY_TCP4) => 12,
+            (V2CMD_PROXY, FAMILY_TCP6) => 36,
+            _ => self.fuse("Unsupported version/command/family")?
+        };
+
+        if addr_len < min_len {
+            self.fuse("Unexpected address length received")?
+        }
+
+        self.read_upto(SIZE_HEADER + addr_len as usize)?;
+        Ok(())
+    }
+
+    fn fuse<T>(&mut self, message: &'static str) -> IoResult<T> {
+        eprintln!("Bailing out of proxy protocol handler: {}", message);
+        self.fused = true;
+        Err(Error::new(ErrorKind::InvalidData, message))
+    }
+
+    pub fn advance(&mut self) -> IoResult<()> {
+        if self.fused {
+            self.fuse("(repeat call)")?
+        }
+
+        self.advance_stream()
+    }
+    pub fn new(stream : TcpStream) -> IoResult<Self> {
+        let mut ret = ProxyState {
+            stream        : stream,
+            fused         : false,
+            buf           : vec![],
+            position      : 0,
+        };
+        ret.advance()?;
+        Ok(ret)
+    }
+}
+
+struct ProxyProtocol {
+    listener      : TcpListener,
+    pub local_address : String,
+}
+
+impl ProxyProtocol {
+    fn new(addr: &str)-> IoResult<Self> {
+        TcpListener::bind(addr).map(|listener| {
+                let local_address = match listener.local_addr() {
+                    Ok(local_address) => local_address.to_string(),
+                    Err(_) => "error".to_string(),
+                };
+                ProxyProtocol {listener, local_address}
+        })
+    }
+}
+
+impl SyncListener for ProxyProtocol {
+    fn accept(&self) -> IoResult<(Box<SyncStream>, Box<ToString>, Box<ToString>)> {
+        let (stream, peer_address) = self.listener.accept()?;
+        let local_address = stream.local_addr().map(|addr| addr.to_string())?;
+        let peer_address = peer_address.to_string();
```

Comment:
You should already strip the HAproxy protocol here so you can return the address found in there.